### ISSUE TITLE
Datetimepicker not rendering properly

### ISSUE THREAD

**Issue by 8vius** [thoughtbot/administrate#1408]
My datetimepicker isn't rendering properly, it shows somewhat collapsed like this
<img width="858" alt="Screen Shot 2019-08-27 at 2 09 13 PM" src="https://user-images.githubusercontent.com/671929/63800704-553a0b00-c8d4-11e9-972f-4a7a16fb2b2a.png">

Has anyone else had this issue?

**Comment by lukaVarga** [comment]
I'm also facing the same issue. There's a JS error in the console:
```
Uncaught TypeError: $(...).datetimepicker is not a function
```

So it looks like administrate is using some sort of jquery datetimepicker library that is missing in the dependency tree. Have you found a solution for it, @8vius?

**Comment by pablobm** [comment]
Fields of type `Field::DateTime` and `Field::Time` are rendered with that picker by default. A default install of Administrate renders it correctly. Something in your installations may be conflicting with it. Can you provide more information? Is there some other JS in your application that may be causing the problem? What version of Rails are you running?

**Comment by lukaVarga** [comment]
@pablobm I'm running Rails 6.0.2. In terms of JS, I have jquery-rails (4.3.5) and turbolinks 5.2.1 but nothing else really. 
For me, `datepicker` method is not defined on the jquery selection - if I do `$('body').datepicker`, it returns `undefined` so it looks as if I have to manually add the datepicker library in order to make it work.

I am running administrate under a namespace, though, (so under `/manager`, not under the default `/admin`). Could that be related? 
Please let me know if I can provide any other info that could be helpful

**Comment by pablobm** [comment]
Hm, I have no idea of what it could be. Here are a few more questions, let's see if we can get to the bottom of it.

What do you get if you run `$('body').datetimepicker`? (Slightly different from what you pasted)

I think that @8vius's problem was different. He got something rendering, so the picker must have ran surely? @lukaVarga, does anything render when you click on a time/datetime field?

What's the name of the field and what options are you passing to it on the dashboard?

Is this a new Rails 6 site, or was it upgraded from a previous version? Does it use Sprockets? Which version?

**Comment by lukaVarga** [comment]
tl;dr: If you change the base layout for Administrate, don't use `defer: true` for JS and if you use Bootstrap, make sure you add `<%= render "stylesheet" %>` after `stylesheet_link_tag`, or to manually re-import `datetime_picker` in your `application.css.sass` **after** you import bootstrap.

@pablobm my bad, I wrote here `.datepicker` but did, in fact, use `.datetimepicker` like in the code. I tried it again - basically, on the initial load, the `$('body').datetimepicker` returns `undefined`. If I then navigated through the app (so the next page will have been served via turbolinks), the `$('foo').datetimepicker` returned a valid function. 

Now, it appears to be a timing issue. At this point, I'm trying out things while writing this comment. I tried disabling turbolinks, didn't work. I then compared my view layout (as I had adjusted it) to the one generated by administrate. I noticed administrate does not provide a `javascript_include_tag` or a `stylesheet_link_tag`, both of which I have added. In my `javascript_include_tag`, I have been using the `defer: true` option and it appears this is what is causing the issue. This made the console error go away, however, I got the same rendered output as OP, so a completely broken datepicker.

At this point, the relevant part of the `head` in my layout looked like this:
```
  <%= render "stylesheet" %>
  <%= csrf_meta_tags %>
  <%= javascript_include_tag 'application', 'data-turbolinks-track': 'reload', defer: false %>
  <%= stylesheet_link_tag 'application', media: 'all', 'data-turbolinks-track': 'reload' %>
```

Just for the fun of it, I decided to switch the order a bit, to see if it's actually conflicting with CSS. Turns out that was the reason for the issue I (and I would assume OP's as well) have experienced with the broken render of the datepicker.

I changed the relevant order in the head to:
```
  <%= csrf_meta_tags %>
  <%= javascript_include_tag 'application', 'data-turbolinks-track': 'reload', defer: false %>
  <%= stylesheet_link_tag 'application', media: 'all', 'data-turbolinks-track': 'reload' %>
  <%= render "stylesheet" %>
```

Now, everything is working as expected :) 

The culprit appears to be Bootstrap that I am using, which I am importing in my `application.css.sass` file. (`gem 'bootstrap', '~> 4.3.1'`). As I don't want to actually change the import of the stylesheets within my HTML (as some CSS classes now get overwritten), I manually imported `datetime_picker` within my `application.css.sass`:
```
@import "bootstrap"
// ... other app related imports
@import "datetime_picker"
```

Voila. 
It's not exactly a bug on administrate's part, rather a nasty piece of conflict between the `datetime_picker_rails` that Administrate is using, and Bootstrap. However, considering Bootstrap is quite popular and considering it's probably not uncommon to have to change the default administrate layout to add some additional JS and CSS, I would suggest 2 notes be added in the readme:
1) Do not use the `defer: true` flag when importing JS files
2) If the project uses Bootstrap 4 (not sure about lower versions), make sure to either manually import `datetime_picker` after the `bootstrap` import within the `application.css.sass`, or to first add app CSS and then to `render 'stylesheet'` in the layout

**Comment by pablobm** [comment]
@lukaVarga - Thank you for that additional information. I'm still a bit confused as to why the loading of the stylesheet would affect this JavaScript, when they should be completely independent. Perhaps something is adding JS in a `content_for :stylesheet` block, which is then loaded with `render "stylesheet"`?

I'm going to close this for now. If other people experience this problem, or something similar, please let us know and we can reopen.
